pkg/dyninst/gosym/symtab.go

```diff
@@ -1180,3 +1193,86 @@ func stringFromOffsetPtr(data []byte, offset int) *string {
 	result := string(data[offset:end])
 	return &result
 }
+
+// FuncPCResolver resolves program counters within a single function to source
+// line numbers. It is optimized for resolving multiple PCs in increasing order.
+type FuncPCResolver struct {
+	funcPC uint64
+	lt     *lineTable
+
+	stepper *stepper
+}
+
+// newFuncResolver builds a FuncPCResolver for the function containing the given
+// program counter (any pc inside the function is valid). Returns nil if the PC
+// could not be resolved to a function.
+func newFuncPCResolver(lt *lineTable, funcPC uint64) *FuncPCResolver {
+	function := lt.getGoFunctionByPC(funcPC)
+	if function == nil {
+		return nil
+	}
+
+	funcInfo, err := lt.funcInfo(function.idx)
+	if err != nil {
+		return nil
+	}
+
+	pcLine, found := funcInfo.pcln()
+	if !found {
+		return nil
+	}
+
+	offset := lt.pcTab[0] + int(pcLine)
+	if offset >= len(lt.data) {
+		return nil
+	}
+	cursor := lt.data[offset:]
+
+	startPC, found := funcInfo.entryPC()
+	if !found {
+		return nil
+	}
+
+	stepper := newStepper(cursor, startPC, -1, lt.quantum)
+	return &FuncPCResolver{
+		funcPC:  funcPC,
+		lt:      lt,
+		stepper: stepper,
+	}
+}
+
+// PCToLine returns the line number corresponding to the given program counter.
+// If the PC corresponds to an inlined function, the returned line number
+// corresponds to the source code of that particular function, with no
+// information being returned about the definition file's name.
+//
+// Any PC inside the function is a valid input, but the FuncPCResolver is more
+// efficient if PCToLine is called with monotonically increasing PCs.
+//
+// Returns false if the PC could not be resolved to a line number.
+func (f *FuncPCResolver) PCToLine(pc uint64) (uint32, bool) {
+	if f.stepper.pc > pc {
+		// We cannot step backwards, so re-initialize the FuncPCResolver and
+		// start the iteration from the beginning of the function.
+		newResolver := newFuncPCResolver(f.lt, f.funcPC)
+		if newResolver == nil {
+			// If we managed to create the current FuncPCResolver, we should be
+			// able to re-create it.
+			panic("failed to re-initialize FuncPCResolver")
+		}
```

Comment:
this feels like a wild panic given this is a public method. The caller could have passed a pc that's outside of the range of the function. Just return false.